### PR TITLE
Payment links: shorter, better

### LINKED ISSUE #195: Improve in-app UX for claimed note senders
Currently, payment links that are claimed by a recipient are not quite first class citizens of the UX afterwards.

1. In history we should track the state of payment links something over the lifecycle "Nalin created a $1.00 payment link" -> "Nalin sent dcposch $1.00 via a payment link" once it is claimed by dcposch.
2. The claimer of the payment link (dcposch) should show on the send screen recent transactions, maybe with a little label like "Sent 10m ago via payment link" or something similar.

### LINKED ISSUE #262: Shorten deep links
### Goal

Ansgar: the payment link URL is too long, looks scary.

#### Before

```
https://daimo.onrender.com/link/note/0x58d3A254a61D592BDEbcf1721e4e9263687AAB02#0xe2702055d60c3f39470a0fd322b85576215b5d34679f0c221534655220ee3af8
```

#### After

```
https://daimo.xyz/link/claim/123712#JibG5hFvh8a5QzB7nvZ
```

- Replace ephemeralOwner address with incrementing ID
- Replace private key with a shorter seed. Above, 14 bytes = 112 bits
- Base58-encode the seed

### LINKED ISSUE #411: Reclaim payment link from other devices
if you're the sender you should allows be shown the reclaim link button even if you don't have its private key

### LINKED ISSUE #441: Claimed payment link: show who it was from
## Issue

You can't see who you got this $2 from:

image

## Proposed fix

Some

Regular inbound transfer

> From **ansgar** (future: link to https://github.com/daimo-eth/daimo/issues/440 )

Claimed payment link

> From **payment link**
> Sender: **dcposch** (future: link to https://github.com/daimo-eth/daimo/issues/440 )

### LINKED ISSUE #496: Transaction History: Show payment links in friendlier way
Annotate links to "Alice's link", "Link to John" etc. as in Michelle's designs

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
Payment link improvements: https://streamable.com/gs1cch

Closes #262
Closes #195 
Closes #441
Closes #496
Closes #411

TODO: make a backcompat plan and test through any backcompat edge cases better

### CODE REVIEW

File: apps/daimo-mobile/src/logic/note.ts [hunk 1]
```diff
@@ -1,19 +1,30 @@
-import { getNoteClaimSignature } from "@daimo/common";
+import {
+  getNoteClaimSignature,
+  getNoteClaimSignatureFromSeed,
+} from "@daimo/common";
 import { useEffect, useState } from "react";
 import { Address, Hex } from "viem";
 
 export function useEphemeralSignature(
```
Comment (dcposch) [inline apps/daimo-mobile/src/logic/note.ts:8]: i'd add a 1 line comment explaining what this does & in particular ephemeralPrivateKey vs seed

File: apps/daimo-mobile/src/sync/sync.ts [hunk 2]
```diff
@@ -301,39 +307,15 @@ function addNamedAccounts(old: EAccount[], found: EAccount[]): EAccount[] {
...
-      txHash: transfer.txHash,
-      blockNumber: transfer.blockNumber,
-      blockHash: transfer.blockHash,
-      logIndex: transfer.logIndex,
-      opHash: transfer.opHash,
-      feeAmount: transfer.feeAmount,
-    });
-  }
-
-  return ret;
```
Comment (dcposch) [inline apps/daimo-mobile/src/sync/sync.ts:338]: If we change the return type of `getAccountHistory`, how do we avoid breaking existing apps?

File: apps/daimo-mobile/src/view/screen/link/NoteScreen.tsx [hunk 3]
```diff
@@ -189,11 +194,8 @@ function NoteDisplayInner({
           <TextBold>Claimed by {getAccountName(noteStatus.claimer!)}</TextBold>
         );
       case "cancelled":
-        if (isOwnSentNote) {
-          return <TextBody>You reclaimed this payment link</TextBody>;
-        } else {
-          return <TextError>Cancelled by sender</TextError>;
-        }
+        assert(isOwnSentNote, "cancelled note not sent by user");
+        return <TextBody>You reclaimed this payment link</TextBody>;
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/link/NoteScreen.tsx:198]: `You cancelled...`

File: apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx [hunk 4]
```diff
@@ -44,57 +44,56 @@ function SendNoteButtonInner({
   account: Account;
   dollars: number;
 }) {
-  const [ephemeralPrivKey] = useState<Hex>(generatePrivateKey);
-  const ephemeralOwner = useMemo(
-    () => privateKeyToAccount(ephemeralPrivKey).address,
-    [ephemeralPrivKey]
-  );
+  const [[noteSeed, noteAddress]] = useState(generateNoteSeedAddress);
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx:47]: 👍

File: apps/daimo-mobile/src/view/screen/link/NoteScreen.tsx (hunk 3 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/link/NoteScreen.tsx:198]: deferring to #571 -- will do it in a separate PR since there's a lot of these

File: apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx [hunk 5]
```diff
@@ -44,57 +44,56 @@ function SendNoteButtonInner({
...
-  ]);
+  const createLinkAccountTransform = (account: Account, pendingOp: OpEvent) => {
+    return {
+      ...transferAccountTransform([
+        {
+          addr: daimoEphemeralNotesAddress,
+          label: AddrLabel.PaymentLink,
+        } as EAccount,
+      ])(account, pendingOp),
+      nextNoteSeq: account.nextNoteSeq + 1,
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx:61]: 👍

File: apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx [hunk 6]
```diff
@@ -44,57 +44,56 @@ function SendNoteButtonInner({
...
       nonceMetadata: nonce.metadata.toHex(),
+      noteStatus: {
+        link: {
+          type: "notev2",
+          sender: account.name,
+          dollars: `${dollars}`,
+          seq: account.nextNoteSeq,
+          seed: noteSeed,
+        },
+        status: "pending",
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx:89]: use enum

File: apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx [hunk 7]
```diff
@@ -44,57 +44,56 @@ function SendNoteButtonInner({
...
+        dollars: `${dollars}`,
+        ephemeralOwner: noteAddress,
+        seq: account.nextNoteSeq,
+      },
     },
-    accountTransform: (account, pendingOp) => {
-      const newAccount = addPendingNote(account, pendingOp);
-      return {
-        ...newAccount,
-        pendingNotes: [
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx:85]: happy we're getting rid of `pendingNotes`

File: apps/daimo-mobile/src/view/shared/HistoryList.tsx [hunk 8]
```diff
@@ -137,7 +140,7 @@ export function HistoryListSwipe({
         }
         return (
           <TransferRow
-            transfer={item.t}
+            displayOp={item.op}
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/shared/HistoryList.tsx:143]: rename to DisplayOpRow

Comment (dcposch) [reply apps/daimo-mobile/src/view/shared/HistoryList.tsx:143]: or HistoryRow  your call

File: apps/daimo-mobile/src/view/shared/HistoryList.tsx [hunk 9]
```diff
@@ -156,25 +159,39 @@ function HeaderRow({ title }: { title: string }) {
   );
 }
 
+function getFromTo(op: DisplayOpEvent): [Address, Address] {
+  if (op.type === "transfer") {
+    return [getAddress(op.from), getAddress(op.to)];
+  } else {
+    if (op.noteStatus.claimer?.addr === op.noteStatus.sender.addr) {
+      // Self-transfer via payment link shows up as two payment link transfers
+      return [getAddress(op.from), getAddress(op.to)];
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/shared/HistoryList.tsx:168]: why do we have to call `getAddress` on op.from but not on op.noteStatus.sender.addr

both should already be Address i think

File: apps/daimo-web/src/app/link/[[...slug]]/AppOrWalletCTA.tsx [hunk 10]
```diff
@@ -227,6 +224,22 @@ async function linkStatusToAction(
...
+        selfAddress,
+        urlHash
+      );
+
+      return {
+        wagmiPrep: {
+          ...daimoEphemeralNotesConfig,
+          functionName: "claimNote" as const,
+          args: [ephemeralOwner, signature] as const,
+        },
```
Comment (dcposch) [inline apps/daimo-web/src/app/link/[[...slug]]/AppOrWalletCTA.tsx:240]: 👍

File: packages/daimo-api/src/api/getAccountHistory.ts [hunk 11]
```diff
@@ -81,6 +84,8 @@ export async function getAccountHistory(
   const lastBlockTimestamp = lastBlk.timestamp;
   const lastBalance = await coinIndexer.getBalanceAt(address, lastBlock);
 
+  const nextNoteSeq = noteIndexer.getNextSeq(address);
```
Comment (dcposch) [inline packages/daimo-api/src/api/getAccountHistory.ts:87]: 👍

File: packages/daimo-api/src/api/getLinkStatus.ts [hunk 12]
```diff
@@ -87,6 +88,26 @@ export async function getLinkStatus(
...
+      }
+      const ret = noteIndexer.getNoteStatusBySeq(sender.addr, link.seq);
+      if (ret == null) {
+        const pending: DaimoNoteStatus = {
+          status: "pending",
+          ephemeralOwner: undefined,
+          link,
+          sender,
+          seq: link.seq,
+          dollars: link.dollars,
```
Comment (dcposch) [inline packages/daimo-api/src/api/getLinkStatus.ts:104]: 👍

File: packages/daimo-api/src/api/getLinkStatus.ts [hunk 13]
```diff
@@ -87,6 +88,26 @@ export async function getLinkStatus(
...
+        };
+        return pending;
+      }
+      return ret;
```
Comment (dcposch) [inline packages/daimo-api/src/api/getLinkStatus.ts:108]: add a check here that logs an error if `ret.seq !== link.seq`

File: packages/daimo-api/src/contract/noteIndexer.ts [hunk 14]
```diff
@@ -51,45 +60,58 @@ export class NoteIndexer {
         where block_num >= $1
         and block_num <= $2
         and chain_id = $3
+        order by (block_num, tx_idx, log_idx) asc
```
Comment (dcposch) [inline packages/daimo-api/src/contract/noteIndexer.ts:63]: 👍

Comment (dcposch) [reply packages/daimo-api/src/contract/noteIndexer.ts:63]: note index interacts poorly w reorgs, but i don't think that's a blocker currently.

File: packages/daimo-common/package.json [hunk 15]
```diff
@@ -25,6 +25,7 @@
   "dependencies": {
     "@daimo/contract": "*",
     "@noble/curves": "^1.2.0",
+    "@scure/base": "^1.1.3",
```
Comment (dcposch) [inline packages/daimo-common/package.json:28]: 👍

File: packages/daimo-common/src/daimoLink.ts [hunk 16]
```diff
@@ -138,32 +158,49 @@ function parseDaimoLinkInner(link: string): DaimoLink | null {
...
+        return {
+          type: "note",
+          previewSender,
+          previewDollars,
+          ephemeralOwner,
+          ephemeralPrivateKey,
+        };
       } else return null;
+    }
+    case "n": {
```
Comment (dcposch) [inline packages/daimo-common/src/daimoLink.ts:183]: maybe `claim` for readability

comparable: `https://peanut.to/claim#?c=8453&v=v4&i=38&p=<key>`

File: packages/daimo-common/test/note.test.ts [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import assert from "node:assert";
+import test from "node:test";
+import { keccak256, verifyMessage } from "viem";
+
+import {
+  generateNoteSeedAddress,
+  getNoteClaimSignatureFromSeed,
+} from "../src/note.ts";
+
```
Comment (dcposch) [inline packages/daimo-common/test/note.test.ts:9]: recommend adding unit test coverage for new link format and backcompat from old to new

**Review by dcposch — COMMENTED** [review]
LGTM

the only thing i'm sketched by is the `nextNoteSeq`

three ideas:
- dont store in Account
- add to notes contract. it would be useful if `sendUserOp` rpc returned an optimistic response (which in this case could include the claimed, guaranteed-unique seqno)
- just use base58-encoded random uint40 / not in sync at all

last case only adds 6 characters to URL, removes code

File: packages/daimo-common/src/daimoLink.ts (hunk 16 above)
Comment (nalinbhardwaj) [reply packages/daimo-common/src/daimoLink.ts:183]: kinda like the shorter link though, we can also make `account` -> `u` and `request` -> `r` in future like reddit, which one do you prefer?

Comment (dcposch) [reply packages/daimo-common/src/daimoLink.ts:183]: sounds good, go for it

File: apps/daimo-mobile/src/logic/note.ts (hunk 1 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/logic/note.ts:12]: done

File: apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx (hunk 6 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/send/SendNoteButton.tsx]: done

File: apps/daimo-mobile/src/view/shared/HistoryList.tsx (hunk 8 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/shared/HistoryList.tsx:144]: done DisplayOpRow

File: packages/daimo-api/src/api/getLinkStatus.ts (hunk 13 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/api/getLinkStatus.ts:109]: deleted seq

File: packages/daimo-api/src/contract/noteIndexer.ts (hunk 14 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/contract/noteIndexer.ts:70]: seq deleted

File: packages/daimo-common/test/note.test.ts (hunk 17 above)
Comment (nalinbhardwaj) [reply packages/daimo-common/test/note.test.ts:9]: done in daimoLink.test.ts

**Comment by nalinbhardwaj** [comment]
> just use base58-encoded random uint40 / not in sync at all

replaced `seq` by `id`: `id` = first 5 char of base58 ephemeralOwner address

File: apps/daimo-mobile/src/view/shared/HistoryList.tsx (hunk 9 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/shared/HistoryList.tsx]: it used to be getAddress before so I was keeping it, but yeah I don't see a reason for it to be now, removed all getAddress 👍

**Review by dcposch — APPROVED** [review]
🚀 looks great to me

**Comment by nalinbhardwaj** [comment]
# Backcompatibility test

Tested by cloning repo twice, running `daimo-api @ nibnalin/pl` in one, `daimo-mobile @ master` in another on two simulators: back and forth few (old style) payment links, everything seems to work fine. Sync seems to work correctly.

Turn off `daimo-mobile @ master`, run `daimo-mobile @ nibnalin/pl`, migration works correctly, on first sync new style of Display Ops works correctly.

Note screen generates new note link style -> works correctly. (i.e. before the last commit on this branch).

Now, turn off new note link style and turn back to old version -> works correctly. (i.e. turn on the last commit on this branch). When we enable new style in a week, we just need to revert this commit and everything should be in a working state.

Separately, the web claim flow works with both old link style and new link style in browser.

 I think this is adequate testing to push this.
